diff --git a/browser/components/tabbrowser/content/tabs.js b/browser/components/tabbrowser/content/tabs.js
index 8aeb244ffca9f48661805f5b7d860b5896055562..d5ac2cf891ec928bec4f95fd5ff0ee8d91294f78 100644
--- a/browser/components/tabbrowser/content/tabs.js
+++ b/browser/components/tabbrowser/content/tabs.js
@@ -94,7 +94,7 @@
       };
       this.arrowScrollbox._canScrollToElement = element => {
         if (isTab(element)) {
-          return !element.pinned || !this.hasAttribute("positionpinnedtabs");
+          return !element.hasAttribute("zen-essential") || !this.hasAttribute("positionpinnedtabs");
         }
         return true;
       };
@@ -362,7 +362,7 @@
       // and we're not hitting the scroll buttons.
       if (
         event.button != 0 ||
-        event.target != this.arrowScrollbox ||
+        event.target != document.getElementById("zen-browser-tabs-wrapper") ||
         event.composedTarget.localName == "toolbarbutton"
       ) {
         return;
@@ -411,6 +411,7 @@
             // Reset the "ignored click" flag
             target._ignoredCloseButtonClicks = false;
           }
+          gZenUIManager.saveScrollbarState();
         }
 
         /* Protects from close-tab-button errant doubleclick:
@@ -455,8 +456,8 @@
           let winUtils = window.windowUtils;
           let endOfTab =
             winUtils.getBoundsWithoutFlushing(lastTab)[
-              (this.verticalMode && "bottom") ||
-                (this.#rtlMode ? "left" : "right")
+            (this.verticalMode && "bottom") ||
+            (this.#rtlMode ? "left" : "right")
             ];
           if (
             (this.verticalMode && event.clientY > endOfTab) ||
@@ -659,7 +660,7 @@
       if (this.#isContainerVerticalPinnedExpanded(tab)) {
         // In expanded vertical mode, the max number of pinned tabs per row is dynamic
         // Set this before adjusting dragged tab's position
-        let pinnedTabs = this.visibleTabs.slice(0, gBrowser.pinnedTabCount);
+        let pinnedTabs = this.visibleTabs.slice(0, gBrowser._numVisiblePinTabs);
         let tabsPerRow = 0;
         let position = 0;
         for (let pinnedTab of pinnedTabs) {
@@ -859,6 +860,9 @@
       }
 
       let draggedTab = event.dataTransfer.mozGetDataAt(TAB_DROP_TYPE, 0);
+      if (draggedTab && effects === "move") {
+        gZenPinnedTabManager.applyDragoverClass(event, draggedTab);
+      }
       if (
         (effects == "move" || effects == "copy") &&
         this == draggedTab.container &&
@@ -972,6 +976,14 @@
 
       this._tabDropIndicator.hidden = true;
       event.stopPropagation();
+      if (draggedTab && dropEffect == "move") {
+        let moved = gZenPinnedTabManager.moveToAnotherTabContainerIfNecessary(event, draggedTab);
+
+        if (moved) {
+          this._finishMoveTogetherSelectedTabs(draggedTab);
+          return;
+        }
+      }
       if (draggedTab && dropEffect == "copy") {
         // copy the dropped tab (wherever it's from)
         let newIndex = this._getDropIndex(event);
@@ -1010,8 +1022,8 @@
           }
         } else {
           let pinned = draggedTab.pinned;
-          let numPinned = gBrowser.pinnedTabCount;
-          let tabs = this.visibleTabs.slice(
+          let numPinned = gBrowser._numVisiblePinTabs;
+          let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab")).slice(
             pinned ? 0 : numPinned,
             pinned ? numPinned : undefined
           );
@@ -1090,7 +1102,7 @@
               let postTransitionCleanup = () => {
                 tab.removeAttribute("tabdrop-samewindow");
 
-                this._finishAnimateTabMove();
+                this._finishAnimateTabMove(true);
                 if (dropIndex !== false) {
                   gBrowser.moveTabTo(tab, dropIndex);
                   if (!directionForward) {
@@ -1100,7 +1112,7 @@
 
                 gBrowser.syncThrobberAnimations(tab);
               };
-              if (gReduceMotion) {
+              if (gReduceMotion || true) {
                 postTransitionCleanup();
               } else {
                 let onTransitionEnd = transitionendEvent => {
@@ -1191,7 +1203,7 @@
         let links;
         try {
           links = browserDragAndDrop.dropLinks(event, true);
-        } catch (ex) {}
+        } catch (ex) { }
 
         if (!links || links.length === 0) {
           return;
@@ -1263,7 +1275,8 @@
       if (
         dt.mozUserCancelled ||
         dt.dropEffect != "none" ||
-        this._isCustomizing
+        this._isCustomizing ||
+        draggedTab.pinned
       ) {
         delete draggedTab._dragData;
         return;
@@ -1346,7 +1359,7 @@
       var left = Math.min(
         Math.max(
           event.screenX * ourCssToDesktopScale -
-            draggedTab._dragData.offsetX * screenCssToDesktopScale,
+          draggedTab._dragData.offsetX * screenCssToDesktopScale,
           availX
         ),
         availX + availWidth - winWidth
@@ -1354,7 +1367,7 @@
       var top = Math.min(
         Math.max(
           event.screenY * ourCssToDesktopScale -
-            draggedTab._dragData.offsetY * screenCssToDesktopScale,
+          draggedTab._dragData.offsetY * screenCssToDesktopScale,
           availY
         ),
         availY + availHeight - winHeight
@@ -1467,7 +1480,7 @@
       // Normal tab title is used also in the permanent private browsing mode.
       const l10nId =
         PrivateBrowsingUtils.isWindowPrivate(window) &&
-        !Services.prefs.getBoolPref("browser.privatebrowsing.autostart")
+          !Services.prefs.getBoolPref("browser.privatebrowsing.autostart")
           ? "tabbrowser-empty-private-tab-title"
           : "tabbrowser-empty-tab-title";
       return gBrowser.tabLocalization.formatValueSync(l10nId);
@@ -1512,9 +1525,18 @@
       }
 
       this.#allTabs = [
-        ...this.verticalPinnedTabsContainer.children,
+        ...document.getElementById("zen-essentials-container").children, ...this.verticalPinnedTabsContainer.children,
         ...children,
       ];
+      for (let i = 0; i < this.#allTabs.length; i++) {
+        // add glance tabs (tabs inside tabs) to the list
+        const glanceTab = this.#allTabs[i].querySelector("tab[zen-glance-tab]");
+        if (glanceTab) {
+          // insert right after the parent tab
+          this.#allTabs.splice(i + 1, 0, glanceTab);
+          i++;
+        }
+      }
       return this.#allTabs;
     }
 
@@ -1593,6 +1615,7 @@
       }
 
       this.#focusableItems = [
+        ...document.getElementById("zen-essentials-container").children,
         ...verticalPinnedTabsContainer.children,
         ...focusableItems,
       ];
@@ -1617,8 +1640,8 @@
     #isContainerVerticalPinnedExpanded(tab) {
       return (
         this.verticalMode &&
-        tab.hasAttribute("pinned") &&
-        this.hasAttribute("expanded")
+        (tab.hasAttribute("zen-essential")) &&
+        (this.hasAttribute("expanded") || document.documentElement.hasAttribute("zen-sidebar-expanded"))
       );
     }
 
@@ -1684,7 +1707,7 @@
       // The height we got may be incorrect if a flush is pending so re-check it after
       // a flush completes.
       window
-        .promiseDocumentFlushed(() => {})
+        .promiseDocumentFlushed(() => { })
         .then(
           () => {
             height =
@@ -1816,7 +1839,7 @@
           let rect = ele => {
             return window.windowUtils.getBoundsWithoutFlushing(ele);
           };
-          let tab = this.visibleTabs[gBrowser.pinnedTabCount];
+          let tab = this.visibleTabs[gBrowser._numVisiblePinTabs];
           if (tab && rect(tab).width <= this._tabClipWidth) {
             this.setAttribute("closebuttons", "activetab");
           } else {
@@ -1832,6 +1855,7 @@
         this.arrowScrollbox.ensureElementIsVisible(selectedTab, aInstant);
       }
 
+      gZenCompactModeManager.flashSidebarIfNecessary(aInstant);
       selectedTab._notselectedsinceload = false;
     }
 
@@ -1843,7 +1867,7 @@
         return;
       }
 
-      let tabs = this.visibleTabs;
+      let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab"));
       if (!tabs.length) {
         return;
       }
@@ -1879,7 +1903,7 @@
         if (isEndTab && !this._hasTabTempMaxWidth) {
           return;
         }
-        let numPinned = gBrowser.pinnedTabCount;
+        let numPinned = gBrowser._numVisiblePinTabs;
         // Force tabs to stay the same width, unless we're closing the last tab,
         // which case we need to let them expand just enough so that the overall
         // tabbar width is the same.
@@ -1894,7 +1918,7 @@
         let tabsToReset = [];
         for (let i = numPinned; i < tabs.length; i++) {
           let tab = tabs[i];
-          tab.style.setProperty("max-width", aTabWidth, "important");
+          //tab.style.setProperty("max-width", aTabWidth, "important");
           if (!isEndTab) {
             // keep tabs the same width
             tab.style.transition = "none";
@@ -1904,7 +1928,7 @@
 
         if (tabsToReset.length) {
           window
-            .promiseDocumentFlushed(() => {})
+            .promiseDocumentFlushed(() => { })
             .then(() => {
               window.requestAnimationFrame(() => {
                 for (let tab of tabsToReset) {
@@ -1965,11 +1989,11 @@
       );
       let numPinned = gBrowser.pinnedTabCount;
 
-      if (gBrowser.pinnedTabCount !== verticalTabsContainer.children.length) {
-        let tabs = this.visibleTabs;
+      if (gBrowser.pinnedTabCount !== (verticalTabsContainer.children.length + document.getElementById("zen-essentials-container").children.length)) {
+        let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab"));
         for (let i = 0; i < numPinned; i++) {
           tabs[i].style.marginInlineStart = "";
-          verticalTabsContainer.appendChild(tabs[i]);
+          tabs[i].hasAttribute("zen-essential") ? document.getElementById("zen-essentials-container").appendChild(tabs[i]) : verticalTabsContainer.appendChild(tabs[i]);
         }
       }
 
@@ -1992,8 +2016,8 @@
     }
 
     _positionPinnedTabs() {
-      let tabs = this.visibleTabs;
-      let numPinned = gBrowser.pinnedTabCount;
+      let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab"));
+      let numPinned = gBrowser._numVisiblePinTabs;
       let absPositionHorizontalTabs =
         this.overflowing && tabs.length > numPinned && numPinned > 0;
 
@@ -2074,7 +2098,7 @@
         return;
       }
 
-      let tabs = this.visibleTabs.slice(0, gBrowser.pinnedTabCount);
+      let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab")).slice(0, gBrowser._numVisiblePinTabs);
 
       let directionX = screenX > dragData.animLastScreenX;
       let directionY = screenY > dragData.animLastScreenY;
@@ -2257,9 +2281,9 @@
       }
 
       let pinned = draggedTab.pinned;
-      let numPinned = gBrowser.pinnedTabCount;
-      let tabs = this.visibleTabs.slice(
-        pinned ? 0 : numPinned,
+      let numPinned = gBrowser._numVisiblePinTabs;
+      let tabs = this.visibleTabs.filter(tab => !tab.hasAttribute("zen-glance-tab")).slice(
+        pinned ? gBrowser._numZenEssentials : numPinned,
         pinned ? numPinned : undefined
       );
 
@@ -2367,8 +2391,8 @@
       };
       let moveOverThreshold = gBrowser._tabGroupsEnabled
         ? Services.prefs.getIntPref(
-            "browser.tabs.dragdrop.moveOverThresholdPercent"
-          ) / 100
+          "browser.tabs.dragdrop.moveOverThresholdPercent"
+        ) / 100
         : 0.5;
       moveOverThreshold = Math.min(1, Math.max(0, moveOverThreshold));
       let newIndex = getDragOverIndex(moveOverThreshold);
@@ -2420,7 +2444,7 @@
         if (
           groupDropIndex in this.allTabs &&
           this.allTabs[groupDropIndex] ==
-            this.allTabs[groupDropIndex].group?.tabs.at(-1)
+          this.allTabs[groupDropIndex].group?.tabs.at(-1)
         ) {
           dragData.groupDropIndex = groupDropIndex;
           dragData.groupDropAction = GROUP_DROP_ACTION_APPEND;
@@ -2502,8 +2526,9 @@
       );
     }
 
-    _finishAnimateTabMove() {
-      if (!this.hasAttribute("movingtab")) {
+    _finishAnimateTabMove(always = false) {
+      gZenPinnedTabManager.removeTabContainersDragoverClass();
+      if (!this.hasAttribute("movingtab") && !always) {
         return;
       }
 
@@ -2552,7 +2577,7 @@
 
         let movingTabSize =
           movingTab.getBoundingClientRect()[
-            this.verticalMode ? "height" : "width"
+          this.verticalMode ? "height" : "width"
           ];
         let shift = (movingTabNewIndex - movingTabOldIndex) * movingTabSize;
 
@@ -2659,18 +2684,17 @@
           let translatePos =
             (this.#rtlMode ? -1 : 1) *
             t._moveTogetherSelectedTabsData.translatePos;
-          t.style.transform = `translate${
-            this.verticalMode ? "Y" : "X"
-          }(${translatePos}px)`;
+          t.style.transform = `translate${this.verticalMode ? "Y" : "X"
+            }(${translatePos}px)`;
         }
       }
 
       function newIndex(aTab, index) {
         // Don't allow mixing pinned and unpinned tabs.
         if (aTab.pinned) {
-          return Math.min(index, gBrowser.pinnedTabCount - 1);
+          return Math.min(index, gBrowser._numVisiblePinTabs - 1);
         }
-        return Math.max(index, gBrowser.pinnedTabCount);
+        return Math.max(index, gBrowser._numVisiblePinTabs);
       }
     }
 
@@ -2754,7 +2778,7 @@
     }
 
     _notifyBackgroundTab(aTab) {
-      if (aTab.pinned || !aTab.visible || !this.overflowing) {
+      if (aTab.hasAttribute("zen-essential") || !aTab.visible || !this.overflowing) {
         return;
       }
 
@@ -2772,12 +2796,14 @@
               selectedTab = {
                 left: selectedTab.left,
                 right: selectedTab.right,
+                top: selectedTab.top,
+                bottom: selectedTab.bottom,
               };
             }
             return [
               this._lastTabToScrollIntoView,
               this.arrowScrollbox.scrollClientRect,
-              { left: lastTabRect.left, right: lastTabRect.right },
+              lastTabRect,
               selectedTab,
             ];
           })
@@ -2794,8 +2820,11 @@
             delete this._lastTabToScrollIntoView;
             // Is the new tab already completely visible?
             if (
-              scrollRect.left <= tabRect.left &&
-              tabRect.right <= scrollRect.right
+              this.verticalMode
+                ? scrollRect.top <= tabRect.top &&
+                tabRect.bottom <= scrollRect.bottom
+                : scrollRect.left <= tabRect.left &&
+                tabRect.right <= scrollRect.right
             ) {
               return;
             }
@@ -2803,21 +2832,29 @@
             if (this.arrowScrollbox.smoothScroll) {
               // Can we make both the new tab and the selected tab completely visible?
               if (
-                !selectedRect ||
-                Math.max(
-                  tabRect.right - selectedRect.left,
-                  selectedRect.right - tabRect.left
-                ) <= scrollRect.width
+                !selectedRect || (this.verticalMode
+                  ? Math.max(
+                    tabRect.bottom - selectedRect.top,
+                    selectedRect.bottom - tabRect.top
+                  ) <= scrollRect.height
+                  : Math.max(
+                    tabRect.right - selectedRect.left,
+                    selectedRect.right - tabRect.left
+                  ) <= scrollRect.width)
               ) {
                 this.arrowScrollbox.ensureElementIsVisible(tabToScrollIntoView);
                 return;
               }
 
-              this.arrowScrollbox.scrollByPixels(
-                this.#rtlMode
-                  ? selectedRect.right - scrollRect.right
-                  : selectedRect.left - scrollRect.left
-              );
+              let scrollPixels;
+              if (this.verticalMode) {
+                scrollPixels = tabRect.top - selectedRect.top;
+              } else if (this.#rtlMode) {
+                scrollPixels = selectedRect.right - scrollRect.right;
+              } else {
+                scrollPixels = selectedRect.left - scrollRect.left;
+              }
+              this.arrowScrollbox.scrollByPixels(scrollPixels);
             }
 
             if (!this._animateElement.hasAttribute("highlight")) {
@@ -2904,7 +2941,7 @@
           sourceNode.localName == "tab" &&
           sourceNode.ownerGlobal.isChromeWindow &&
           sourceNode.ownerDocument.documentElement.getAttribute("windowtype") ==
-            "navigator:browser" &&
+          "navigator:browser" &&
           sourceNode.ownerGlobal.gBrowser.tabContainer == sourceNode.container
         ) {
           // Do not allow transfering a private tab to a non-private window
