diff --git a/tools/signing/macos/mach_commands.py b/tools/signing/macos/mach_commands.py
index a513ad723805459c194d27b42dac68e9babba468..2988a1b73ce10edf4a155a1bb6bae2945a8929e0 100644
--- a/tools/signing/macos/mach_commands.py
+++ b/tools/signing/macos/mach_commands.py
@@ -342,6 +342,9 @@ def macos_sign(
     cs_reset_cmd = ["find", app, "-exec", "codesign", "--remove-signature", "{}", ";"]
     run(command_context, cs_reset_cmd, capture_output=not verbose_arg)
 
+    # Move ./embedded.provisionprofile to [app]/Contents/embedded.provisionprofile
+    run(command_context, ["mv", "./embedded.provisionprofile", os.path.join(app, "Contents")], capture_output=not verbose_arg)
+
     if use_rcodesign_arg is True:
         sign_with_rcodesign(
             command_context,
