diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index 9c3bb513537f1efc44788fc3efb9f7561eca9c0e..dcada8742c5ea147d2378357c86c9a8b7dc639db 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -81,10 +81,13 @@ option(
 )
 set_config("MOZ_INCLUDE_SOURCE_INFO", True, when="MOZ_INCLUDE_SOURCE_INFO")
 
+option(env="ZEN_FIREFOX_VERSION", help="Set Zen version of Firefox", nargs=1)
+set_config("ZEN_FIREFOX_VERSION", depends_if("ZEN_FIREFOX_VERSION")(lambda v: v[0]))
+
 option(
     "--with-distribution-id",
     nargs=1,
-    default="org.mozilla",
+    default="app.zen-browser",
     help="Set distribution-specific id",
 )
 set_config("MOZ_DISTRIBUTION_ID", depends("--with-distribution-id")(lambda v: v[0]))
@@ -869,9 +872,9 @@ set_define("MOZ_AV1", av1)
 option("--disable-jxl", help="Disable jxl image support")
 
 
-@depends("--disable-jxl", milestone.is_nightly)
-def jxl(value, is_nightly):
-    if is_nightly and value:
+@depends("--disable-jxl")
+def jxl(value):
+    if value:
         return True
 
 
@@ -2012,7 +2015,7 @@ set_define("A11Y_LOG", True, when=a11y_log)
 # ==============================================================
 @depends(milestone)
 def require_signing(milestone):
-    return milestone.is_release_or_beta and not milestone.is_esr
+    return False
 
 
 option(
