diff --git a/toolkit/content/widgets/browser-custom-element.js b/toolkit/content/widgets/browser-custom-element.js
index a531577a29212cdb9e745b88e17fe4177be76635..ae976ecef71d0d659fa6a1d6fea60f5270073544 100644
--- a/toolkit/content/widgets/browser-custom-element.js
+++ b/toolkit/content/widgets/browser-custom-element.js
@@ -556,11 +556,11 @@
       if (!this.browsingContext) {
         return;
       }
-      this.browsingContext.isActive = val;
+      this.browsingContext.isActive = val || this.zenModeActive;
       if (this.isRemoteBrowser) {
         let remoteTab = this.frameLoader?.remoteTab;
         if (remoteTab) {
-          remoteTab.renderLayers = val;
+          remoteTab.renderLayers = val || this.zenModeActive;
         }
       }
     }
